File: c/src/simulation.c
// simulation.c

#include <stdio.h>
#include <math.h>
#include "simulation.h"

// Function to subtract two vectors
Vector vector_sub(Vector a, Vector b) {
    Vector result;
    result.x = a.x - b.x;
    result.y = a.y - b.y;
    return result;
}

// Function to add two vectors
Vector vector_add(Vector a, Vector b) {
    Vector result;
    result.x = a.x + b.x;
    result.y = a.y + b.y;
    return result;
}

// Function to multiply a vector by a scalar
Vector vector_mul(Vector v, double scalar) {
    Vector result;
    result.x = v.x * scalar;
    result.y = v.y * scalar;
    return result;
}

// Function to divide a vector by a scalar
Vector vector_div(Vector v, double scalar) {
    Vector result;
    result.x = v.x / scalar;
    result.y = v.y / scalar;
    return result;
}

// Function to calculate the magnitude of a vector
double vector_magnitude(Vector v) {
    return sqrt(v.x * v.x + v.y * v.y);
}

// Function to compute forces on all particles
void compute_forces(Particle particles[], Vector forces[], int n) {
    for (int i = 0; i < n; i++) {
        forces[i] = (Vector){0.0, 0.0};  // Initialize force to zero
        for (int j = 0; j < n; j++) {
            if (i != j) {
                Vector displacement = vector_sub(particles[j].position, particles[i].position);
                double distance = vector_magnitude(displacement);
                if (distance != 0) {
                    double force_magnitude = G * particles[i].mass * particles[j].mass / (distance * distance);
                    Vector force_vector = vector_mul(vector_div(displacement, distance), force_magnitude);
                    forces[i] = vector_add(forces[i], force_vector);
                }
            }
        }
    }
}

// Function to update positions and velocities of particles
void update_positions_and_velocities(Particle particles[], Vector forces[], int n, double dt) {
    for (int i = 0; i < n; i++) {
        Vector acceleration = vector_div(forces[i], particles[i].mass);
        particles[i].velocity = vector_add(particles[i].velocity, vector_mul(acceleration, dt));
        particles[i].position = vector_add(particles[i].position, vector_mul(particles[i].velocity, dt));
    }
}

// Function to simulate the motion of particles
void simulate(Particle particles[], int num_particles, int num_steps, double dt) {
    Vector forces[num_particles];
    for (int step = 0; step < num_steps; step++) {
        compute_forces(particles, forces, num_particles);
        update_positions_and_velocities(particles, forces, num_particles, dt);
    }
}

File: python/src/pure_python/simulation.py
from typing import List

# Constants
G: float = 6.67430e-11  # Gravitational constant
DT: float = 1e5  # Time step (seconds)
NUM_STEPS: int = 1000  # Run simulation


class Vector:
    def __init__(self, x: float = 0, y: float = 0):
        self.x: float = x
        self.y: float = y

    def __sub__(self, other: 'Vector') -> 'Vector':
        return Vector(self.x - other.x, self.y - other.y)

    def __add__(self, other: 'Vector') -> 'Vector':
        return Vector(self.x + other.x, self.y + other.y)

    def __mul__(self, scalar: float) -> 'Vector':
        return Vector(self.x * scalar, self.y * scalar)

    def __truediv__(self, scalar: float) -> 'Vector':
        return Vector(self.x / scalar, self.y / scalar)

    def magnitude(self) -> float:
        return (self.x ** 2 + self.y ** 2) ** 0.5

    def __repr__(self) -> str:
        return f"Vector(x={self.x}, y={self.y})"


class Particle:
    def __init__(self, position: Vector, velocity: Vector, mass: float):
        self.position: Vector = position
        self.velocity: Vector = velocity
        self.mass: float = mass


def compute_forces(particles):
    forces = [Vector() for _ in particles]
    for _i, particle_i in enumerate(particles):
        for _j, particle_j in enumerate(particles):
            if _i != _j:
                displacement = particle_j.position - particle_i.position
                distance = displacement.magnitude()
                if distance != 0:
                    force_magnitude = G * particle_i.mass * particle_j.mass / distance ** 2
                    force_vector = displacement / distance * force_magnitude
                    forces[_i] += force_vector

    return forces


def update_positions_and_velocities(particles: List[Particle], forces: List[Vector], dt: float):
    for _, _particle in enumerate(particles):
        acceleration = forces[_] / _particle.mass
        _particle.velocity += acceleration * dt
        _particle.position += _particle.velocity * dt


def simulate(particles: List[Particle], num_steps: int = NUM_STEPS, dt: float = DT):
    for _ in range(num_steps):
        forces: List[Vector] = compute_forces(particles)
        update_positions_and_velocities(particles, forces, dt)
